### PR TITLE
Feature: Infer schema type automatically.

### PR DESCRIPTION
**Pull request by mohammad0-0ahmad**
# Tasks to be implemented :
- [x] Infer schema type automatically from document definition.
- [x] Provide TS utility to be able to extract schema type from schema instance to be able to reuse the type as separated type "InferSchemaType".
- [x] Improve Model.create function type to show schema properties in suggestions dialog.
- [x]  Adding "statics"  property in schema options  to support auto type inference for model statics methods.
- [x]  Adding "methods"  property in schema options  to support auto type inference for document instance methods.
- [x] Support custom type key "typeKey" to inference schema type.
- [x] Improve Model new function "document constructor" type to show schema properties in suggestions dialog.
- [x] Improve Model.insertMany function type to show schema properties in suggestions dialog.
- [x]  Adding "query"  property in schema options  to support auto type inference for query helpers methods.

# Status:
- [x]  Ready to review.
- [x]  Docs is updated.
- [x]  Ready to merge.

Do you have any suggestions or ideas to be implemented in this PR, please leave a comment :slightly_smiling_face:

### Get it early:
You can install mongoose including this feature by running:
```shell
 npm i Automattic/mongoose#pull/11563/head
```
# Screenshots :
- Auto schema type inference & create function:

Untitled

- InferSchemaType utility & supporting custom type key:

Untitled

### CODE REVIEW

**Comment by Uzlopak** [comment]
@mohammad0-0ahmad 

When do you think this PR will be ready for merge?

**Comment by mohammad0-0ahmad** [comment]
> @mohammad0-0ahmad
> 
> When do you think this PR will be ready for merge?

When it will be reviewed :)

File: test/document.test.js
Comment (Uzlopak) [inline test/document.test.js:11243]: please remove your iniitials in the title of the unit test describe

```suggestion
describe('Check if instance functions that is supplied in schema option is availabe', function() {
```

File: test/model.test.js
Comment (Uzlopak) [inline test/model.test.js:8494]: please remove your iniitials in the title of the unit test describe

```suggestion
describe('Check if statics functions that is supplied in schema option is availabe', function() {
```

File: test/query.test.js
Comment (Uzlopak) [inline test/query.test.js:3965]: will this even work? We are in a callback.

```suggestion
      if (error instanceof Error) {
        return done(error);
      }
```

File: test/query.test.js
Comment (Uzlopak) [inline test/query.test.js:3967]:
```suggestion
        if (error instanceof Error) {
          return done(error);
        }
```

File: test/types/document.test.ts
Comment (Uzlopak) [inline test/types/document.test.ts:182]: nitpick
I am not fan of "blocks" like these
Just simple
// Document-Method-tests 

Also the other "blocks" should be changed...

File: test/types/models.test.ts
Comment (Uzlopak) [inline test/types/models.test.ts:231]: You use everywhere a prefix of 'M0_0a'. I woud prefer that you remove this and in the inferschematype.d.ts you write good header like

```
/*
 * @author Mohammad Ahmad (mohammad0-0ahmad) your@email.com
 */

```
If @vkarpov15  allows it, you could add yourself to the contributors list in the package.json.

**Review by Uzlopak — COMMENTED** [review]
Not easy to digest  :D

Made some comments.

**Comment by mohammad0-0ahmad** [comment]
@Uzlopak 
Thanks for reviewing and following the entire process.
I've resolved the latest comments, and I am looking for @vkarpov15 review.
Best regards! 🙂

**Comment by Uzlopak** [comment]
@vkarpov15 
I think this should be part of v6.3

**Comment by Uzlopak** [comment]
Maybe we should talk about documenting https://github.com/Automattic/mongoose/blob/master/docs/typescript.md

**Comment by mohammad0-0ahmad** [comment]
Hello @vkarpov15!
Would you like to review this?

**Comment by uncle-ara** [comment]
when will typing become available for general use?

**Comment by mohammad0-0ahmad** [comment]
> when will typing become available for general use?

I am waiting for @vkarpov15  review.

**Comment by Uzlopak** [comment]
@mohammad0-0ahmad 

I hope you are still motivated. Issue is that complex typescript PR are always leaving an unease feeling in the stomach. So dont take it personally. :)

Can you check please if we need some documentation changes? 

https://github.com/Automattic/mongoose/tree/master/docs/typescript

**Comment by mohammad0-0ahmad** [comment]
> @mohammad0-0ahmad
> 
> I hope you are still motivated. Issue is that complex typescript PR are always leaving an unease feeling in the stomach. So dont take it personally. :)
> 
> Can you check please if we need some documentation changes?
> 
> https://github.com/Automattic/mongoose/tree/master/docs/typescript

I am still motivated, and I understand that, but I've provided some JS docs for the type, and maybe to explain the unclear things in this PR.
We need absolutely to make some documentation and maybe update docs on the website, but I would like to wait until get review response in case something will be not acceptable.

File: types/index.d.ts
Comment (vkarpov15) [inline types/index.d.ts:305]: Why `UnpackedIntersection<>` here? `T` shouldn't be an array as far as I can tell. This may be contributing to the performance issues.

File: types/index.d.ts
Comment (vkarpov15) [inline types/index.d.ts:306]: Can you still call `create()` with arbitrary data? There's a good reason why we have `AnyKeys<T> | AnyObject` as the param type here: you can technically call `create()` with any object.

File: types/schemaoptions.d.ts
Comment (vkarpov15) [inline types/schemaoptions.d.ts:10]: I really like the custom type key support, this is great :+1:

File: types/index.d.ts
Comment (vkarpov15) [inline types/index.d.ts:134]: Why drop this declaration? This would be backwards breaking

**Review by vkarpov15 — COMMENTED** [review]
I love the type key support and auto typed schemas. But my big concerns are 1) performance, 2) backwards btreaking changes. Our TS benchmark shows before this PR:

```
Run node ./benchmarks/typescript
simple instantiations: 35805
simple memory used: 152972.7
simple check time: 3.089
simple total time: 4.831999999999999
```

After this PR:

```
Run node ./benchmarks/typescript
simple instantiations: 227139
simple memory used: 191137.1
simple check time: 3.8959999999999995
simple total time: 5.3
```

After spending over 6 months fixing the performance of our TypeScript types, I do not want to merge something that undoes all the progress we made in #10349.

**Comment by mohammad0-0ahmad** [comment]
> 

@vkarpov15  Thanks for replying.
I will answer the rest of your comments.
Would you like to provide me how can I benchmark the typing?
I couldn't find info about that in the contributing docs. 
I've tested with --diagnostics flag, but I got different results.

File: types/index.d.ts
Comment (mohammad0-0ahmad) [reply types/index.d.ts:134]: It's not backwards breaking by far, but it affects TS performance actually, so I've restored it.

File: types/schemaoptions.d.ts
Comment (mohammad0-0ahmad) [reply types/schemaoptions.d.ts:10]: Thanks

File: types/index.d.ts
Comment (mohammad0-0ahmad) [reply types/index.d.ts:313]: Yes, it is still working with  arbitrary data. Related test: 
https://github.com/Automattic/mongoose/pull/11563/files#diff-bb912bdd6eed058fe45695e309bce313922f26da17bf41827fee94b5c55e2996R224

File: types/index.d.ts
Comment (mohammad0-0ahmad) [reply types/index.d.ts:312]: This is not affect the performance as far as I can tell, and it is an array right now.

File: types/document.d.ts
Comment (mohammad0-0ahmad) [reply types/document.d.ts:196]: This just to support Model type with any possible schema type.

**Comment by mohammad0-0ahmad** [comment]
> I love the type key support and auto typed schemas. But my big concerns are 1) performance, 2) backwards btreaking changes. Our TS benchmark shows before this PR:
> 
> ```
> Run node ./benchmarks/typescript
> simple instantiations: 35805
> simple memory used: 152972.7
> simple check time: 3.089
> simple total time: 4.831999999999999
> ```
> 
> After this PR:
> 
> ```
> Run node ./benchmarks/typescript
> simple instantiations: 227139
> simple memory used: 191137.1
> simple check time: 3.8959999999999995
> simple total time: 5.3
> ```
> 
> After spending over 6 months fixing the performance of our TypeScript types, I do not want to merge something that undoes all the progress we made in #10349.

@vkarpov15 
I understand your concerns, and I've always thought to avoid backwards breaking during the process, but I've not noticed the performance changing.
Anyway, I've got the following result after the latest refactor, and I think It is a good result so far.

```
Run node ./benchmarks/typescript
simple instantiations: 4[6](https://github.com/Automattic/mongoose/runs/6220922425?check_suite_focus=true#step:5:6)468
simple memory used: 15535[7](https://github.com/Automattic/mongoose/runs/6220922425?check_suite_focus=true#step:5:7).6
simple check time: 2.577
simple total time: 4.01
```

**Comment by mohammad0-0ahmad** [comment]
Hello @Uzlopak  @vkarpov15 !
I've updated some docs files, please double-check them before merging.
Best regards.

File: docs/typescript/statics.md
Comment (Uzlopak) [inline docs/typescript/statics.md:27]:
```suggestion
Mongoose does support auto typed static functions that it are supplied in schema options.
```

File: docs/typescript/statics.md
Comment (Uzlopak) [inline docs/typescript/statics.md:28]:
```suggestion
Static functions can be defined by:
```

File: test/document.test.js
Comment (Uzlopak) [inline test/document.test.js:11297]:
```suggestion
describe('Check if instance function that is supplied in schema option is availabe', function() {
```

File: test/model.test.js
Comment (Uzlopak) [inline test/model.test.js:8508]:
```suggestion
describe('Check if statics function that is supplied in schema option is availabe', function() {
```

File: docs/guide.md
Comment (Uzlopak) [inline docs/guide.md:149]:
```suggestion
  // This approach is supported by auto typed instance typescript functionality. There is no need to manually create  separate TS types. 
```

File: docs/guide.md
Comment (Uzlopak) [inline docs/guide.md:184]:
```suggestion
- Add a function property to the second argument of the schema-constructor (`statics`)
```

File: docs/guide.md
Comment (Uzlopak) [inline docs/guide.md:194]:
```suggestion
  // This approach is supported by auto typed instance typescript functionality. There is no need to manually create  separate TS types. 
```

File: docs/guide.md
Comment (Uzlopak) [inline docs/guide.md:228]:
```suggestion
  // This approach is supported by auto typed instance typescript functionality. There is no need to manually create  separate TS types. 
```

File: docs/typescript/query-helpers.md
Comment (Uzlopak) [inline docs/typescript/query-helpers.md:17]:
```suggestion
In TypeScript, Mongoose does support manually typed and automatically typed Query Helpers.
```

File: docs/typescript/query-helpers.md
Comment (Uzlopak) [inline docs/typescript/query-helpers.md:64]:
```suggestion
Mongoose does support auto typed Query Helpers that it are supplied in schema options.
```

File: docs/typescript/schemas.md
Comment (Uzlopak) [inline docs/typescript/schemas.md:49]:
```suggestion
// InferSchemaType will determine the type as follows: 
```

File: docs/typescript/schemas.md
Comment (Uzlopak) [inline docs/typescript/schemas.md:34]:
```suggestion
// No need to define TS interface any more.
```

**Review by Uzlopak — CHANGES_REQUESTED** [review]
I am not a native english speaker. I think my requested changes make sense. But maybe some proficient speaker can provide better wording.

**Comment by mohammad0-0ahmad** [comment]
> I am not a native english speaker. I think my requested changes make sense. But maybe some proficient speaker can provide better wording.

Done, Thanks for the review @Uzlopak

File: test/types/create.test.ts
Comment (vkarpov15) [inline test/types/create.test.ts:12]: This change is a bit concerning. We should at least support passing ObjectIds as strings in `create()`, you shouldn't need the `new Types.ObjectId()` here.

File: test/model.test.js
Comment (vkarpov15) [inline test/model.test.js:8508]: Typo "available"

**Review by vkarpov15 — COMMENTED** [review]
I had one minor comment about `create()` with strings for ObjectIds, but otherwise this looks great :+1:

Can I change the branch to 6.4 so we can ship this with our next minor release?

File: test/types/create.test.ts
Comment (mohammad0-0ahmad) [reply test/types/create.test.ts:12]: It supports passing object as string, but then it needs to use  type assertion to workaround ts error like so:
```
Test.create({ _id: '0'.repeat(24) as unknown as Types.ObjectId, name: 'test' }).then((doc: ITest) => console.log(doc.name));
```
Or like the following but doc type will be string instead of object id.
```
Test.create({ _id: '0'.repeat(24) as unknown as Types.ObjectId, name: 'test' }).then((doc) => console.log(doc.name));

```
else, ITest interface can be as following:

```
interface ITest extends Document {
  _id?: Types.ObjectId | string;
  name?: string;
}
```

File: test/model.test.js
Comment (mohammad0-0ahmad) [reply test/model.test.js]: > Typo "available"

done..

**Comment by mohammad0-0ahmad** [comment]
> I had one minor comment about `create()` with strings for ObjectIds, but otherwise this looks great 👍
> 
> Can I change the branch to 6.4 so we can ship this with our next minor release?

Sure :)

**Comment by Uzlopak** [comment]
Hi, 

can you please fix the merge conflicts? Sry for the inconvenience

File: test/types/schema.test.ts
Comment (Uzlopak) [inline test/types/schema.test.ts:381]: We are missing Schema.Types.Decimal128

`
    decimal1: Schema.Types.Decimal128,`

File: test/types/schema.test.ts
Comment (Uzlopak) [inline test/types/schema.test.ts:344]:
```typescript
    decimal1: Schema.Types.Decimal128,
    decimal2: 'Decimal128',
```

File: types/inferschematype.d.ts
Comment (Uzlopak) [inline types/inferschematype.d.ts:140]: I guess you would need to do 

```typescript
                PathValueType extends 'Decimal128' | typeof Schema.Types.Decimal128 ? Schema.Types.Decimal128 :
```

File: test/types/create.test.ts
Comment (Uzlopak) [reply test/types/create.test.ts:12]: I think this is a serious regression, breaking user space, which must be avoided. The workarounds are not feasable, because mongoose internally casts the _id string to a ObjectId. So we have to ensure if _id is defined as  ObjectId that _id is defined as `_id?: Types.ObjectId | string`

File: types/index.d.ts
Comment (Uzlopak) [inline types/index.d.ts:422]: I assume the same thing as create. I guess _id is now also strictly typed as ObjectId

**Comment by Uzlopak** [comment]
Please merge master into your PR. 

I applied #11886, which according to my findings should enable you to add 
`PathValueType extends 'Decimal128' | typeof Schema.Types.Decimal128 ? Schema.Types.Decimal128 :`
to inferschematype.d.ts successfully as the discriminator schemaName is set to proper values.

File: types/inferschematype.d.ts
Comment (mohammad0-0ahmad) [reply types/inferschematype.d.ts:141]: @Uzlopak 
Nice you got it, I wonder if 'decimal128' also available to  use as well.

File: test/types/create.test.ts
Comment (mohammad0-0ahmad) [reply test/types/create.test.ts:12]: @Uzlopak  @vkarpov15 
I would like just to clarify some things here:
1- This specific example use enforced doc type which means It's not auto typed doc, Therefore I think the user should be the responsible to provide the correct type here. 
2- mongoose allows defining _id path as a number  as mentioned here https://mongoosejs.com/docs/guide.html#_id.
Therefore, I think it is wrong to make _id always as 'string' | 'object'  which will not be correct in this case.
3- we can make every path that is defined as 'Objected' in schema definition as string | ObjectId in auto typed schema, but then it will be ObjectId | string for all paths not just _id, so is that correct then ?

**Comment by mohammad0-0ahmad** [comment]
> Created a PR against your fork/branch
> 
> [mohammad0-0ahmad-forks#38](https://github.com/mohammad0-0ahmad-forks/mongoose/pull/38)
> 
> There were some issues regarding populate, because you changed Model from Interface to type.
> 
> So this line would have broken, because Model would not extend from Model in strict mode. https://github.com/mohammad0-0ahmad-forks/mongoose/blob/7cebaa50a964c061ea805670e7913c8f4c860359/test/types/populate.test.ts#L75
> 
> I fixed the typings of insertMany and create.

That line doesn't exist in master branch, anyway I've copied the mentioned line, but I did not get any TS error.
Can you please describe the problem more?
Capture

**Comment by Uzlopak** [comment]
I fixed the merge errors in the patch PR.

I can not pinpoint which change resulted into the issue of not being able to use any Model. But it fixed it. Maybe when I patched the Require_id type. Idk. 

Please review and hopefully merge

**Comment by mohammad0-0ahmad** [comment]
> I fixed the merge errors in the patch PR.
> 
> I can not pinpoint which change resulted into the issue of not being able to use any Model. But it fixed it. Maybe when I patched the Require_id type. Idk.
> 
> Please review and hopefully merge

I've reviewed changes briefly and merged.

**Review by vkarpov15 — APPROVED** [review]
Nice, thanks :+1: We'll merge this into the 6.4 branch. Thanks for all your hard work on this.

**Comment by mohammad0-0ahmad** [comment]
Thank you @Uzlopak  @vkarpov15  as well.

**Comment by magyarb** [comment]
Hi, and thanks for all the work done here.

I see that [mongoose-tsgen](https://www.npmjs.com/package/mongoose-tsgen#known-issues) is no longer updated due to this PR.

Will it be possible to export these inferred types as separate files? It would be useful to use these types or interfaces elsewhere, for example on the frontend, where mongoose is not installed, or converting them to json schemas, to perform input validation on both the frontend and backend.

**Comment by mohammad0-0ahmad** [comment]
It might be great if you can make availabe internally @francescov1 or at least mention that ts-gen can generate types files in mongoose docs.
What do you think @vkarpov15 @Uzlopak @AbdelrahmanHafez ?

**Comment by vkarpov15** [comment]
@francescov1 I'd say continue to support mongoose-tsgen for now :+1:

**Comment by francescov1** [comment]
@vkarpov15 Will do 👍

File: types/index.d.ts
Comment (Starrah) [inline types/index.d.ts:105]: Sorry that I'm not excellent in Typescript, but what does `U extends any` here mean? 

As far as I know, everything extends any? so the contiditional clause `(T & { _id: Types.ObjectId })` is always used, and the clause `T & Required<{ _id: U }>` is never used. 

Thus, `_id` will be always infered with type `ObjectId`, even if I have defined another type such as `String` in the Schema definition.

Comment (Starrah) [reply types/index.d.ts:105]: For example:
```ts
import {model, Schema} from "mongoose";
const schema_with_string_id = new Schema({_id: String, nickname: String})
const theModel = model('test', schema_with_string_id)
const obj = new theModel()
const theId = obj._id
```
The type of variable `theId` will be `string | Types.ObjectId`, rather than `string`. 
Is this designed behaviour for some reason I don't know, or just a bug?

Comment (Starrah) [reply types/index.d.ts:105]: @mohammad0-0ahmad Sorry for bothering you!

Comment (mohammad0-0ahmad) [reply types/index.d.ts:105]: Hello @Starrah !
Firstly, you are not bothering, and you are totally right.
I think this meant to be  something like:
```TS
  export type Require_id<T> = T extends { _id?: infer U }
    ? IfEquals<U, any> extends true
      ? (T & { _id: Types.ObjectId })
      : T & Required<{ _id: U }>
    : T & { _id: Types.ObjectId };
```
Thanks for your review, I will make sure to refactor this.
What are your thoughts @Uzlopak  about that?

Comment (Uzlopak) [reply types/index.d.ts:105]: I think I did this to check if T is not never.

Comment (Starrah) [reply types/index.d.ts:105]: > I think I did this to check if T is not never.

First, does you mean you did that to check if `{_id: }` in T is not never, rather than check T?
Because according to my experiment, 
```ts
import {Require_id, Types} from "mongoose";

type Required_id_old<T> = T extends { _id?: any } ? (T & { _id: T['_id'] }) : (T & { _id: Types.ObjectId });
type Required_id_new<T> = T extends { _id?: infer U }
    ? U extends any
        ? (T & { _id: Types.ObjectId })
        : T & Required<{ _id: U }>
    : T & { _id: Types.ObjectId };

type t1_old = Required_id_old<{_id: never}>
type t1_new = Required_id_new<{_id: never}>

type t2_old = Required_id_old<never>
type t2_new = Required_id_new<never>
```
Both `t2_old` and `t2_new` is `never`(which I think is expected behaviour), and for t1, `t1_old` is `{_id: never}` but `t1_new` is `never`(I have no idea which is your expected behaviour).

Then, if my assumption is true and you does mean `check if `{_id: }` in T is not never`, then maybe the two conditional clause should be swapped, as the following:
```ts
type Required_id_new<T> = T extends { _id?: infer U }
    ? U extends any
        ? T & Required<{ _id: U }>
        : (T & { _id: Types.ObjectId })
    : T & { _id: Types.ObjectId };
```

Comment (Starrah) [reply types/index.d.ts:105]: @mohammad0-0ahmad @Uzlopak 
Though with the discussion above, we (or just I) cannot still figure out what the codes here means, specifically, what does the `U extends any` mean.  

**However, I want to point out that the following test case written by me, got failed:**
```ts
// index.test-d.ts
import {model, Schema} from "mongoose";
import {expectType} from "tsd";

const schema_with_string_id = new Schema({_id: String, nickname: String})
const theModel = model('test', schema_with_string_id)
const obj = new theModel()

expectType<string>(obj._id)
```
when testing it with `tsd`, gives the following error: 
```
 index.test-d.ts:8:0
  ✖  8:0  Parameter type string is declared too wide for argument type string & ObjectId.
```
According to @mohammad0-0ahmad :


> Hello @Starrah ! Firstly, you are not bothering, and you are totally right. I think this meant to be something like:
> 
> ```ts
>   export type Require_id<T> = T extends { _id?: infer U }
>     ? IfEquals<U, any> extends true
>       ? (T & { _id: Types.ObjectId })
>       : T & Required<{ _id: U }>
>     : T & { _id: Types.ObjectId };
> ```
> 
> Thanks for your review, I will make sure to refactor this. What are your thoughts @Uzlopak about that?

I think that the test case failure above **indicates a bug**. So, I decide to raise a new issue (#12070) and assign @mohammad0-0ahmad, to report and/or further discuss this bug.

**Comment by olawalejuwonm** [comment]
Does this work for method too?

**Pull request merged**
